src/lib/operators/projection.cpp

```diff
@@ -290,4 +304,53 @@ std::shared_ptr<Table> Projection::dummy_table() {
   return shared_dummy;
 }
 
+/**
+ *  Method to determine PQPColumn that should not be forwarded. In most cases, forwarding a column comes with almost
+ *  zero costs and should be done. However, in cases where the same PQPColumn is also accessed in the expression
+ *  evaluation, the whole column is materialized and cached anyways. In that case, returning the materialized PQPColumn
+ *  instead of forwarding it benefits the following operator which can process a fully materialized column sequentially
+ *  instead of accessing the column indirectly via a position list.
+ */
+std::vector<bool> Projection::_determine_pqp_columns_to_evaluate() const {
+  const auto expression_count = expressions.size();
+
+  // To decide which forwardable columns should be evaluated, we first need to know if they are part of an expression.
+  // As expressions that reference a PQPColumn might be place before or behind the forwardable PQPColumn, we run twice
+  // over the expressions. In a first step, we collect the number of occurences of PQPColumns.
+  auto pqp_column_id_occurences = std::vector<size_t>(left_input_table()->column_count());
+  for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
+    const auto& expression = expressions[column_id];
+    visit_expression(expression, [&](const auto& sub_expression) {
+      if (const auto pqp_column_expression = std::dynamic_pointer_cast<PQPColumnExpression>(sub_expression)) {
+        const auto pqp_column_id = pqp_column_expression->column_id;
+        if (pqp_column_id != INVALID_COLUMN_ID) {
+          ++pqp_column_id_occurences[pqp_column_id];
+        }
+      }
+      return ExpressionVisitation::VisitArguments;
+    });
+  }
+
+  // In case a forwarded segment is dictionary encoded (e.g., TPC-H Q1), aggregate optimizations for dictionary
+  // encoding could potentially render evaluation instead of forwarding here disadvantegeous. Thus, we check if the
+  // column occurs more often than the threshold. As of now, a threshold of 2 is fine (i.e., we evaluate a forwardable
+  // column as soon as it is referenced again), which worked fine for TPC-H.
+  // Depending on changes on other operators, it might be beneficial to forward dictionary columns more often. But for
+  // encodings such as LZ4, a threshold of two should be best.
+  constexpr auto THRESHOLD = 2;
+
+  auto pqp_columns_to_evaluate = std::vector<bool>(expression_count, false);
+  for (auto column_id = ColumnID{0}; column_id < expression_count; ++column_id) {
+    const auto& expression = expressions[column_id];
+    if (const auto pqp_column_expression = std::dynamic_pointer_cast<PQPColumnExpression>(expression)) {
+      const auto pqp_column_id = pqp_column_expression->column_id;
+      if (pqp_column_id_occurences[pqp_column_id] >= THRESHOLD) {
+        pqp_columns_to_evaluate[column_id] = true;
+      }
+    }
+  }
+
+  return pqp_columns_to_evaluate;
+}
+
```

Comment:
This is a lot of code to essentially just extract all PQPColumnExpressions that are not top-level output expressions (i.e., have to be evaluated).

Do you have any numbers that suggest that having a threshold is beneficial? As of now, this appears to be dead code.

Reply:
I started the PR on sidon and this PR was not really faster than the master (for dictionary, for LZ4 is was still a win).
I'll think about it.

Reply:
Removed it. Neglectable on sidon and nobody will ever re-evaluate this threshold anyways.